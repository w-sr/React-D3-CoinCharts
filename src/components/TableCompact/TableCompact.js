import React, { Component } from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import currencyFormatter from "currency-formatter";

import BigAmount from "../BigAmount";
import Flex from "../Flex";
import { PROPTYPES } from "../../constants";
import { animation, border, color, fontFamily, fontSize, fontWeight, size } from "../../styles/constants";
import arrowIcon from "./assets/arrow-icon.png";

const DOWN_ARROW_DESCRIPTION = "Arrow pointing downwards";
const UP_ARROW_DESCRIPTION = "Arrow pointing upwards";

const StyledTableCompact = styled(Flex)`
  box-sizing: border-box;
  color: ${color.slateDark};
  padding: ${size.large} 0;
  height: 140px;
`;

const Dropdown = styled.select`
  -webkit-appearance: none;
  background-position: right 8px center;
  background: url(${arrowIcon}) right 8px center / 8px no-repeat ${color.white};
  border-image: initial;
  border-radius: ${border.radius};
  border: ${border.border}
  color: ${color.slateDark};
  cursor: pointer;
  flex: 1 1 0%;
  font-family: ${fontFamily.regular};
  font-size: ${fontSize.small};
  font-weight: ${fontWeight.medium};
  max-width: 100px;
  outline: none;
  padding: 8px 22px 8px 8px;
  text-transform: uppercase;
  transition: ${animation.default};
`;

const PriceDetails = styled.div.attrs({
  color: props => (props.positive ? color.positive : color.slateDark),
  transform: props => (props.positive ? "rotate(0deg)" : "rotate(180deg)"),
})`
  color: ${props => props.color};
  font-size: ${fontSize.small};
  font-weight: ${fontWeight.medium};

  svg {
    fill: ${props => props.color};
    height: 11px;
    margin-right: 4px;
    transform: ${props => props.transform};
    width: 11px;
  }
`;

class TableCompact extends Component {
  static renderSVGArrow(isDirectionUpward) {
    return (
      <svg height="12" width="12" viewBox="8135 12432 15 15" version="1.1" xmlns="http://www.w3.org/2000/svg">
        <title>Arrow</title>
        <desc>{isDirectionUpward ? UP_ARROW_DESCRIPTION : DOWN_ARROW_DESCRIPTION}</desc>
        <path fillRule="evenodd" d="M8149.72 12439.36l-7.36-7.36-7.36 7.36 1 1 5.64-5.64v11.82h1.4v-11.87l5.68 5.69z" />
      </svg>
    );
  }

  constructor(props) {
    super(props);

    // Bind event-handlers
    this.handleOnSelectChange = this.handleOnSelectChange.bind(this);
  }

  shouldComponentUpdate(nextProps) {
    const { isLoading } = nextProps;
    return !isLoading;
  }

  handleOnSelectChange(event) {
    const { onCryptocurrencyChange } = this.props;
    onCryptocurrencyChange(event.target.value);
  }

  render() {
    const {
      currency,
      spotPrice,
      priceDifference,
      percentDifference,
      cryptocurrencies,
      selectedCryptocurrency,
    } = this.props;
    const formattedPriceDifference = currencyFormatter.format(Math.abs(priceDifference), { code: currency });
    const formattedPercentDifference = Math.abs(percentDifference).toFixed(2);

    return (
      <StyledTableCompact align="center" justify="space-between">
        <div>
          <BigAmount type="currency" currency={currency} value={spotPrice} />
          <PriceDetails positive={priceDifference > 0}>
            {TableCompact.renderSVGArrow(priceDifference > 0)}
            <span data-testid="price-details">{`${formattedPriceDifference} (${formattedPercentDifference}%)`}</span>
          </PriceDetails>
        </div>
        <Dropdown value={selectedCryptocurrency} onChange={this.handleOnSelectChange}>
          {cryptocurrencies.map(crypto => (
            <option key={crypto.key} value={crypto.key}>
              {crypto.key}
            </option>
          ))}
        </Dropdown>
      </StyledTableCompact>
    );
  }
}

TableCompact.propTypes = {
  isLoading: PropTypes.bool.isRequired,
  percentDifference: PropTypes.number.isRequired,
  priceDifference: PropTypes.number.isRequired,
  spotPrice: PropTypes.number.isRequired,
  currency: PropTypes.string.isRequired,
  cryptocurrencies: PropTypes.arrayOf(
    PropTypes.shape({
      key: PropTypes.string,
    }),
  ).isRequired,
  selectedCryptocurrency: PROPTYPES.CRYPTOCURRENCY.isRequired,
  onCryptocurrencyChange: PropTypes.func.isRequired,
};

export default TableCompact;
