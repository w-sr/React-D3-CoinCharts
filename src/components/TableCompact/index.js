import TableCompact from "./TableCompact";

export default TableCompact;
