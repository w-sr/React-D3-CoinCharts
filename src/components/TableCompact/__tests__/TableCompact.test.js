import React from "react";
import { mount } from "enzyme";
import { render } from "react-testing-library";

import TableCompact from "../";

const TEST_PROPS = {
  onCryptocurrencyChange: jest.fn(),
  selectedCryptocurrency: "BTC",
  cryptocurrencies: [{ key: "BTC" }, { key: "BCH" }, { key: "ETH" }, { key: "LTC" }],
  currency: "USD",
  isLoading: false,
  percentDifference: 9.99,
  priceDifference: 99.99,
  spotPrice: 1999.99,
};

describe("<TableCompact />", () => {
  it("renders without crashing", () => {
    const { container } = render(<TableCompact {...TEST_PROPS} />);
    expect(container.firstChild).not.toBeNull();
    expect(container.firstChild).toMatchSnapshot();
  });

  it("renders negative differences correctly", () => {
    const props = {
      ...TEST_PROPS,
      percentDifference: -9.99,
      priceDifference: -99.99,
    };
    const expected = `$${Math.abs(props.priceDifference)} (${Math.abs(props.percentDifference)}%)`;

    const { getByTestId } = render(<TableCompact {...props} />);
    expect(getByTestId("price-details").textContent).toEqual(expected);
  });

  it("renders positive differences correctly", () => {
    const expected = `$${Math.abs(TEST_PROPS.priceDifference)} (${Math.abs(TEST_PROPS.percentDifference)}%)`;

    const { getByTestId } = render(<TableCompact {...TEST_PROPS} />);
    expect(getByTestId("price-details").textContent).toEqual(expected);
  });

  it("does not rerender when `props.isLoading` is true", () => {
    const { container } = render(<TableCompact {...TEST_PROPS} />);
    const { textContent } = container.firstChild;

    // Component should not rerender
    render(<TableCompact {...TEST_PROPS} percentDifference={1000} isLoading />, { container });
    expect(container.firstChild.textContent).toEqual(textContent);

    // Component should rerender
    render(<TableCompact {...TEST_PROPS} percentDifference={1000} />, { container });
    expect(container.firstChild.textContent).not.toEqual(textContent);
  });

  // TODO: Switch to "react-testing-library" once https://github.com/kentcdodds/react-testing-library/pull/48 is merged
  it("triggers `props.onChange` callback when option is selected", () => {
    const initialSelectedOption = "BTC";
    const selectedOption = "LTC";
    const props = {
      ...TEST_PROPS,
      selectedCryptocurrency: initialSelectedOption,
    };

    const wrapper = mount(<TableCompact {...props} />);
    wrapper.find("select").simulate("change", { target: { value: selectedOption } });
    expect(props.onCryptocurrencyChange.mock.calls[0][0]).toEqual(selectedOption);
  });
});
