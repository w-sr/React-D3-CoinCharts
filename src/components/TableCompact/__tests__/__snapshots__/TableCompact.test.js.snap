// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TableCompact /> renders without crashing 1`] = `
<div
  class="sc-bxivhb kPYCRu sc-htpNat izTaYc"
>
  <div>
    <div
      class="sc-bdVaJa jlxkRb"
    >
      <span
        color="#4E5C6E"
        class="sc-bwzfXH fbCOIX"
      >
        $
      </span>
      <span>
        1,999
      </span>
      <span
        color="#4E5C6E"
        class="sc-bwzfXH fbCOIX"
      >
        .99
      </span>
    </div>
    <div
      color="#61CA00"
      transform="rotate(0deg)"
      class="sc-EHOje dqfVBA"
    >
      <svg
        height="12"
        width="12"
        viewBox="8135 12432 15 15"
        version="1.1"
        xmlns="http://www.w3.org/2000/svg"
      >
        <title>
          Arrow
        </title>
        <desc>
          Arrow pointing upwards
        </desc>
        <path
          fill-rule="evenodd"
          d="M8149.72 12439.36l-7.36-7.36-7.36 7.36 1 1 5.64-5.64v11.82h1.4v-11.87l5.68 5.69z"
        />
      </svg>
      <span
        data-testid="price-details"
      >
        $99.99 (9.99%)
      </span>
    </div>
  </div>
  <select
    class="sc-ifAKCX dKuzaW"
  >
    <option
      value="BTC"
    >
      BTC
    </option>
    <option
      value="BCH"
    >
      BCH
    </option>
    <option
      value="ETH"
    >
      ETH
    </option>
    <option
      value="LTC"
    >
      LTC
    </option>
  </select>
</div>
`;
